### PR TITLE
Andygolay/bridge e2e test framework

### PR DESCRIPTION
**Pull request by andygolay**
# Summary
Currently working off `62c33771d78524ddf71dc7308f5dce0588602077` rev of our Aptos Core fork.

# Testing

1. Run the local Suzuka node and bridge:

```
CELESTIA_LOG_LEVEL=FATAL CARGO_PROFILE=release CARGO_PROFILE_FLAGS=--release  nix develop --extra-experimental-features nix-command --extra-experimental-features flakes --command bash  -c "just bridge native build.setup.eth-local.celestia-local.bridge --keep-tui"
```

2. In `.movement/config.yaml` change the `account` field from
```
f90391c81027f03cdea491ed8b36ffaced26b6df208a9b569e5baf2590eb9b16
```
to 
```
0xA550C18
```
which allows use of the core resource account in bridge setup scripts.

The bridge operator is set to `f90391c81027f03cdea491ed8b36ffaced26b6df208a9b569e5baf2590eb9b16` in the setup, because that's the address derived from the signer key set up in Genesis.

3. Run the Eth -> Movement E2E test against the framework. 
```
DOT_MOVEMENT_PATH="$(pwd)/.movement" cargo test --test bridge_e2e_test_framework test_bridge_transfer_eth_movement_happy_path -- --nocapture --test-threads=1
```
# Outstanding issues
Still need Movement -> Eth direction E2E tests. 

Also we hard-coded `FRAMEWORK_ADDRESS` in some places, e.g. `counterparty_complete_bridge_transfer` that may be a bit of a hack.

### CODE REVIEW

File: protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+		client_framework::MovementClientFramework, event_monitoring::MovementMonitoring, utils::MovementAddress,
+	},
+};
+use bridge_service::types::Amount;
+use bridge_service::types::AssetType;
+use bridge_service::types::BridgeAddress;
+use bridge_service::types::HashLock;
+use bridge_service::types::HashLockPreImage;
+use futures::StreamExt;
+use tracing_subscriber::EnvFilter;
```
Comment (0xmovses) [inline protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs:29]: Could we clean up these imports, make more succinct.

File: protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+	// .await;
+	// println!("res: {res:?}",);
+
+	// let res = test_get_events_by_account_event_handle(
+	// 	&config.movement.mvt_rpc_connection_url(),
+	// 	"0xf90391c81027f03cdea491ed8b36ffaced26b6df208a9b569e5baf2590eb9b16",
+	// 	&event_type,
+	// )
+	// .await;
+	// println!("res: {res:?}",);
```
Comment (0xmovses) [inline protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs:227]: Let's remove all the commented out code. We don't need it for ref anymore.

File: protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+	{
+		let faucet_client = mvt_client_harness.faucet_client.write().unwrap();
+		faucet_client.fund(movement_client_signer_address, 100_000_000).await?;
+	}
+
+	let recipient_privkey = mvt_client_harness.fund_account().await;
+	let recipient_address = MovementAddress(recipient_privkey.address());
+
+	// 1) initialize Eth transfer
+	tracing::info!("Call initiate_transfer on Eth");
```
Comment (0xmovses) [inline protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs:101]: The numbering comments don't continue, no need to number.

File: protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+	// let amount = Amount(AssetType::EthAndWeth((1, 0)));
+	// initiate_eth_bridge_transfer(
+	// 	&config,
+	// 	HarnessEthClient::get_initiator_private_key(&config),
+	// 	mov_recipient,
+	// 	hash_lock,
+	// 	amount,
+	// )
+	// .await
+	// .expect("Failed to initiate bridge transfer");
```
Comment (0xmovses) [inline protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs:180]: Same with comments here, remove them.

File: protocol-units/bridge/service/src/main.rs [hunk 5]
```diff
@@ -26,14 +26,17 @@ async fn main() -> Result<()> {
 
 	// get a matching godfig object
 	let godfig: Godfig<Config, ConfigFile> = Godfig::new(ConfigFile::new(config_file), vec![]);
-	let bridge_config: Config = godfig.try_wait_for_ready().await?;
+	let mut bridge_config: Config = godfig.try_wait_for_ready().await?;
+
+	//bridge_config.movement = MovementConfig::suzuka();
```
Comment (0xmovses) [inline protocol-units/bridge/service/src/main.rs:31]: Remove the commented out line.

**Review by 0xmovses — CHANGES_REQUESTED** [review]
Good to see this passing!

Code cleanup nits, we've been pretty messy in these e2e tests, good to cleanup now before these go in.

**Comment by 0xmovses** [comment]
After this goes in, let's create a PR that will run the e2e test in CI. As currently we are just running integration tests in CI.

**Comment by 0xmovses** [comment]
Not sure if i'm missing step, I followed the instructions in the PR description 

```
stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::system_addresses: ENOT_CORE_RESOURCE_ADDRESS(0x50001): The address/account did not correspond to the core resource address"
}

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::system_addresses: ENOT_CORE_RESOURCE_ADDRESS(0x50001): The address/account did not correspond to the core resource address"
}

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::system_addresses: ENOT_CORE_RESOURCE_ADDRESS(0x50001): The address/account did not correspond to the core resource address"
}

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::system_addresses: ENOT_CORE_RESOURCE_ADDRESS(0x50001): The address/account did not correspond to the core resource address"
}

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::system_addresses: ENOT_CORE_RESOURCE_ADDRESS(0x50001): The address/account did not correspond to the core resource address"
}
```

**Comment by 0xmovses** [comment]
I think I need to update the `config.yaml` `account` field but not sure what it should be update to. 

It's currently set to `f90391c81027f03cdea491ed8b36ffaced26b6df208a9b569e5baf2590eb9b16`

**Comment by andygolay** [comment]
> I think I need to update the `config.yaml` `account` field but not sure what it should be update to.
> 
> It's currently set to `f90391c81027f03cdea491ed8b36ffaced26b6df208a9b569e5baf2590eb9b16`

Oh, yeah thank you, that needs to be set to `0xA550C18` prior to running the E2E test command, and reset each time you start up the local node. I'll update the PR description and add that setup into `bridge_setup_scripts`.

**Comment by 0xmovses** [comment]
With Account updated I am now getting this, am I missing something?
```
stdout: {
  "Result": {
    "transaction_hash": "0xc14d7431b46c168da97956142830024e50918ca942535863e5f57455a6907c04",
    "gas_used": 7,
    "gas_unit_price": 100,
    "sender": "000000000000000000000000000000000000000000000000000000000a550c18",
    "sequence_number": 16,
    "success": true,
    "timestamp_us": 1729634096006996,
    "version": 102,
    "vm_status": "Executed successfully"
  }
}

stderr: Transaction submitted: https://explorer.movementlabs.xyz/txn/0xc14d7431b46c168da97956142830024e50918ca942535863e5f57455a6907c04?network=custom

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::transaction_fee: ECOPY_CAPS_SHOT(0x7): "
}

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::atomic_bridge_configuration: EINVALID_BRIDGE_OPERATOR(0x1): Error code for invalid bridge operator"
}

stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::atomic_bridge_configuration: EINVALID_BRIDGE_OPERATOR(0x1): Error code for invalid bridge operator"
}

stderr: 
stdout: {
  "Error": "Simulation failed with status: Move abort in 0x1::atomic_bridge_configuration: EINVALID_BRIDGE_OPERATOR(0x1): Error code for invalid bridge operator"
}

stderr: 
2024-10-22T21:55:08.275383Z  INFO bridge_e2e_test_framework: Init initiator and counter part test account.
2024-10-22T21:55:08.275406Z  INFO bridge_e2e_test_framework: Use client signer for Mvt and index 2 of config.eth.eth_well_known_account_private_keys array for Eth
2024-10-22T21:55:13.812773Z  INFO bridge_e2e_test_framework: Call initiate_transfer on Eth
base_call_builder: CallBuilder { request: TransactionRequest { from: Some(0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc), to: Some(Call(0x8a791620dd6260079bf849dc5567adc3f2fdc318)), gas_price: None
, max_fee_per_gas: None, max_priority_fee_per_gas: None, max_fee_per_blob_gas: None, gas: None, value: Some(1), input: TransactionInput { input: Some(0x1a8849a40000000000000000000000000000000000
000000000000000000000000000000df550fdc48abe4d0a685ae317459f8b4a6fff34de0cb80bb790386c3746c342ce482bbf23babfaccd31fe29ec5864e65d038a4077c4acf84a4aca482ea1ddfc7), data: None }, nonce: None, chain_
id: None, access_list: None, transaction_type: None, blob_versioned_hashes: None, sidecar: None }, block: Number(Latest), state: None, decoder: "bridge_service::chains::ethereum::types::AtomicBr
idgeInitiator::initiateBridgeTransferCall" }
Sending transaction with gas limit: 10000000000000000
estimated_gas: 204056
Sending transaction with gas: 204056
2024-10-22T21:55:13.822860Z  INFO bridge_e2e_test_framework: Wait for the MVT Locked event.
thread 'test_bridge_transfer_eth_movement_happy_path' panicked at protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs:121:9:
Not a Locked event: Some(Err(OnChainError("MVT get_account_events json convertion error:error decoding response body")))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
FAILED

failures:

failures:
    test_bridge_transfer_eth_movement_happy_path

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 1 filtered out; finished in 30.23s
```

**Comment by andygolay** [comment]
> With Account updated I am now getting this, am I missing something?
> 
> ```
> stdout: {
>   "Result": {
>     "transaction_hash": "0xc14d7431b46c168da97956142830024e50918ca942535863e5f57455a6907c04",
>     "gas_used": 7,
>     "gas_unit_price": 100,
>     "sender": "000000000000000000000000000000000000000000000000000000000a550c18",
>     "sequence_number": 16,
>     "success": true,
>     "timestamp_us": 1729634096006996,
>     "version": 102,
>     "vm_status": "Executed successfully"
>   }
> }
> 
> stderr: Transaction submitted: https://explorer.movementlabs.xyz/txn/0xc14d7431b46c168da97956142830024e50918ca942535863e5f57455a6907c04?network=custom
> 
> stdout: {
>   "Error": "Simulation failed with status: Move abort in 0x1::transaction_fee: ECOPY_CAPS_SHOT(0x7): "
> }
> 
> stdout: {
>   "Error": "Simulation failed with status: Move abort in 0x1::atomic_bridge_configuration: EINVALID_BRIDGE_OPERATOR(0x1): Error code for invalid bridge operator"
> }
> 
> stdout: {
>   "Error": "Simulation failed with status: Move abort in 0x1::atomic_bridge_configuration: EINVALID_BRIDGE_OPERATOR(0x1): Error code for invalid bridge operator"
> }
> 
> stderr: 
> stdout: {
>   "Error": "Simulation failed with status: Move abort in 0x1::atomic_bridge_configuration: EINVALID_BRIDGE_OPERATOR(0x1): Error code for invalid bridge operator"
> }
> 
> stderr: 
> 2024-10-22T21:55:08.275383Z  INFO bridge_e2e_test_framework: Init initiator and counter part test account.
> 2024-10-22T21:55:08.275406Z  INFO bridge_e2e_test_framework: Use client signer for Mvt and index 2 of config.eth.eth_well_known_account_private_keys array for Eth
> 2024-10-22T21:55:13.812773Z  INFO bridge_e2e_test_framework: Call initiate_transfer on Eth
> base_call_builder: CallBuilder { request: TransactionRequest { from: Some(0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc), to: Some(Call(0x8a791620dd6260079bf849dc5567adc3f2fdc318)), gas_price: None
> , max_fee_per_gas: None, max_priority_fee_per_gas: None, max_fee_per_blob_gas: None, gas: None, value: Some(1), input: TransactionInput { input: Some(0x1a8849a40000000000000000000000000000000000
> 000000000000000000000000000000df550fdc48abe4d0a685ae317459f8b4a6fff34de0cb80bb790386c3746c342ce482bbf23babfaccd31fe29ec5864e65d038a4077c4acf84a4aca482ea1ddfc7), data: None }, nonce: None, chain_
> id: None, access_list: None, transaction_type: None, blob_versioned_hashes: None, sidecar: None }, block: Number(Latest), state: None, decoder: "bridge_service::chains::ethereum::types::AtomicBr
> idgeInitiator::initiateBridgeTransferCall" }
> Sending transaction with gas limit: 10000000000000000
> estimated_gas: 204056
> Sending transaction with gas: 204056
> 2024-10-22T21:55:13.822860Z  INFO bridge_e2e_test_framework: Wait for the MVT Locked event.
> thread 'test_bridge_transfer_eth_movement_happy_path' panicked at protocol-units/bridge/integration-tests/tests/bridge_e2e_test_framework.rs:121:9:
> Not a Locked event: Some(Err(OnChainError("MVT get_account_events json convertion error:error decoding response body")))
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> FAILED
> 
> failures:
> 
> failures:
>     test_bridge_transfer_eth_movement_happy_path
> 
> test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 1 filtered out; finished in 30.23s
> ```

Is this for the first time you run the test? Or subsequent times?

And is there some error about the specific thing that's not being decoded, in the bridge logs?

**Comment by 0xmovses** [comment]
When restarting the node, it works. Probably some improvements we can make on the e2e test before it can go into CI. But this is all checking out. Thank you!

**Review by 0xmovses — APPROVED** [review]

**Pull request merged**